File: server-laravel/app/Models/Task.php
<?php

namespace App\Models;

use Rhino\Models\RhinoModel;
use Rhino\Traits\HasAuditTrail;
use App\Models\Project;
use App\Models\User;


class Task extends RhinoModel
{
    use HasAuditTrail;

    // ---------------------------------------------------------------
    // Route Key (see "Route Key" in the Rhino docs)
    // ---------------------------------------------------------------
    // Member endpoints (show/update/destroy/restore/force-delete) match the
    // {id} URL segment against `hash_id` instead of the primary key.
    public static string $routeKey = 'hash_id';

    protected $fillable = [
            'title',
            'description',
            'status',
            'priority',
            'estimated_hours',
            'due_date',
            'project_id',
            'assignee_id',
        ];

    // ---------------------------------------------------------------
    // Validation rules
    // ---------------------------------------------------------------

    // Format rules for all fields (applied on both store and update).
    protected $validationRules = [
            'title' => 'required|string|max:255',
            'description' => 'nullable|string',
            'status' => 'required|string|max:255',
            'priority' => 'required|string|max:255',
            'estimated_hours' => 'nullable|numeric',
            'due_date' => 'nullable|date',
            'project_id' => 'required|integer|exists:projects,id',
            'assignee_id' => 'nullable|integer|exists:users,id',
        ];

    // ---------------------------------------------------------------
    // Query Builder configuration (used by Rhino's GlobalController)
    // ---------------------------------------------------------------

    public static $allowedFilters = [
            'title',
            'status',
            'priority',
        ];
    public static $allowedSorts = [
            'title',
            'status',
            'priority',
            'due_date',
        ];
    public static $defaultSort = 'created_at';

    // ---------------------------------------------------------------
    // Named scopes selectable with ?scope=
    // ---------------------------------------------------------------
    //
    //   ?scope=assignedToMe                              no arguments
    //   ?scope[dueBefore]=2026-12-31                     one declared parameter
    //   ?scope[dueBetween][from]=a&scope[dueBetween][to]=b    named parameters
    //   ?scope[byStatus][status]=todo                    'priority' is optional
    //
    public static $allowedScopes = [
            'assignedToMe',
            'dueBefore' => 'date',
            'dueBetween' => ['from', 'to'],
            'byStatus' => ['params' => ['status', 'priority'], 'optional' => ['priority']],
        ];
    public static $defaultScope = 'active';
    public static $allowedFields = [
            'id',
            'hash_id',
            'title',
            'description',
            'status',
            'priority',
            'estimated_hours',
            'due_date',
            'project_id',
            'assignee_id',
            'created_at',
            'updated_at',
        ];
    public static $allowedIncludes = [
            'project',
            'assignee',
        ];
    // public static $allowedSearch = [];

    // ---------------------------------------------------------------
    // Pagination (uncomment to enable default pagination)
    // ---------------------------------------------------------------
    // public static bool $paginationEnabled = false;
    // protected $perPage = 25;

    // ---------------------------------------------------------------
    // Middleware
    // ---------------------------------------------------------------
    public static array $middlewareActions = ['store' => ['throttle:60,1']];

    // ---------------------------------------------------------------
    // Exclude actions (uncomment to disable specific CRUD endpoints)
    // ---------------------------------------------------------------
    // public static array $exceptActions = [];

    // ---------------------------------------------------------------
    // Hidden columns (add columns to hide from API responses)
    // ---------------------------------------------------------------
    // protected static $additionalHiddenColumns = [];

    // ---------------------------------------------------------------
    // Computed attributes (see "Computed Attributes" in the Rhino docs)
    // ---------------------------------------------------------------

    // OPT-IN per-row values: nothing is evaluated unless the client asks for it
    // by name via ?computed_attributes=comment_count,is_overdue
    public function rhinoRecordComputedAttributes(): array
    {
        return [
            'comment_count' => fn ($record, $user) => $record->comments()->count(),
            'is_overdue' => fn ($record, $user) => $record->due_date !== null
                && $record->status !== 'done'
                && $record->due_date < now(),
        ];
    }

    // COLLECTION-level aggregates: evaluated ONCE per request over the scoped,
    // filtered query. Served by GET /api/{org}/tasks/computed?attributes=...
    public static function rhinoCollectionComputedAttributes(): array
    {
        return [
            'total_count' => fn ($query, $user) => $query->count(),
            'open_tasks_count' => fn ($query, $user) => $query->where('status', '!=', 'done')->count(),
            'done_tasks_count' => fn ($query, $user) => $query->where('status', 'done')->count(),
            'high_priority_count' => fn ($query, $user) => $query->where('priority', 'high')->count(),
            'estimated_hours_total' => fn ($query, $user) => (float) $query->sum('estimated_hours'),
        ];
    }

    // ---------------------------------------------------------------
    // Named scope bodies. Each receives the current user first, then the
    // parameters declared in $allowedScopes, in that order.
    // ---------------------------------------------------------------

    public function scopeActive(\Illuminate\Database\Eloquent\Builder $query, ?\Illuminate\Contracts\Auth\Authenticatable $user): \Illuminate\Database\Eloquent\Builder
    {
        return $query->where('status', '!=', 'done');
    }

    public function scopeAssignedToMe(\Illuminate\Database\Eloquent\Builder $query, ?\Illuminate\Contracts\Auth\Authenticatable $user): \Illuminate\Database\Eloquent\Builder
    {
        if (! $user) {
            return $query->whereRaw('1 = 0'); // fail closed
        }

        return $query->where('assignee_id', $user->id);
    }

    public function scopeDueBefore(\Illuminate\Database\Eloquent\Builder $query, ?\Illuminate\Contracts\Auth\Authenticatable $user, $date): \Illuminate\Database\Eloquent\Builder
    {
        return $query->whereNotNull('due_date')->whereDate('due_date', '<', $date);
    }

    public function scopeDueBetween(\Illuminate\Database\Eloquent\Builder $query, ?\Illuminate\Contracts\Auth\Authenticatable $user, $from, $to): \Illuminate\Database\Eloquent\Builder
    {
        return $query->whereBetween('due_date', [$from, $to]);
    }

    public function scopeByStatus(\Illuminate\Database\Eloquent\Builder $query, ?\Illuminate\Contracts\Auth\Authenticatable $user, $status, $priority = null): \Illuminate\Database\Eloquent\Builder
    {
        $query->where('status', $status);

        if ($priority !== null) {
            $query->where('priority', $priority);
        }

        return $query;
    }

    // ---------------------------------------------------------------
    // Relationships
    // ---------------------------------------------------------------

    public function project(): \Illuminate\Database\Eloquent\Relations\BelongsTo
    {
        return $this->belongsTo(Project::class);
    }

    public function assignee(): \Illuminate\Database\Eloquent\Relations\BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function comments(): \Illuminate\Database\Eloquent\Relations\HasMany
    {
        return $this->hasMany(Comment::class);
    }

    public function labels(): \Illuminate\Database\Eloquent\Relations\BelongsToMany
    {
        return $this->belongsToMany(Label::class, 'task_labels');
    }
}

File: server-rails/app/models/task.rb
# frozen_string_literal: true

class Task < Rhino::RhinoModel
  include Rhino::HasAuditTrail
  include Discard::Model

  rhino_filters :title, :status, :priority
  rhino_sorts :title, :status, :priority, :due_date
  rhino_default_sort "created_at"

  # ---------------------------------------------------------------
  # Named scopes selectable with ?scope=
  # ---------------------------------------------------------------
  #
  #   ?scope=assignedToMe                                  no arguments
  #   ?scope[dueBefore]=2026-12-31                         one declared parameter
  #   ?scope[dueBetween][from]=a&scope[dueBetween][to]=b   named parameters
  #   ?scope[byStatus][status]=todo                        'priority' is optional
  #
  rhino_scopes :active,
               assigned_to_me: { with: lambda { |relation, user|
                 user ? relation.where(assignee_id: user.id) : relation.none
               } },
               due_before: { params: [:date] },
               due_between: { params: %i[from to] },
               by_status: { params: %i[status priority], optional: [:priority] }
  rhino_default_scope :active

  scope :active, -> { where.not(status: "done") }
  scope :due_before, ->(date) { where.not(due_date: nil).where("due_date < ?", date) }
  scope :due_between, ->(from, to) { where(due_date: from..to) }
  scope :by_status, lambda { |status, priority = nil|
    relation = where(status: status)
    priority ? relation.where(priority: priority) : relation
  }
  rhino_fields :id, :hash_id, :title, :description, :status, :priority, :estimated_hours, :due_date, :project_id, :assignee_id, :created_at, :updated_at
  rhino_includes :project, :assignee

  # Route Key: member endpoints match hash_id instead of the numeric id
  # (GET /api/{org}/tasks/{hash_id}; numeric ids no longer match).
  rhino_route_key :hash_id

  validates :title, length: { maximum: 255 }, allow_nil: true
  validates :status, inclusion: { in: %w[todo in_progress in_review done] }, allow_nil: true
  validates :priority, inclusion: { in: %w[low medium high critical] }, allow_nil: true

  # ---------------------------------------------------------------
  # Computed attributes (see "Computed Attributes" in the Rhino docs)
  # ---------------------------------------------------------------

  # OPT-IN per-row values: nothing is evaluated unless the client asks for it
  # by name via ?computed_attributes=comment_count,is_overdue
  #
  # An attribute may also declare PARAMETERS the client supplies, using the
  # same bracket wire form named scopes use:
  #
  #   ?computed_attributes=comment_count,is_overdue      legacy comma list
  #   ?computed_attributes[comment_count]=               bracket, no arguments
  #   ?computed_attributes[is_due_before]=2026-01-01     one declared parameter
  #   ?computed_attributes[is_due_before][date]=...      named parameter
  #
  # Bound arguments arrive after the user, in the order `params:` declares them.
  def rhino_record_computed_attributes
    {
      "comment_count" => ->(record, _user) { record.comments.count },
      "is_overdue" => lambda { |record, _user|
        record.due_date.present? && record.status != "done" && record.due_date < Date.current
      },

      # Parameterised record attribute: is this row due before the
      # client-supplied date?
      "is_due_before" => {
        params: [:date],
        with: ->(record, _user, date) { record.due_date.present? && record.due_date < Date.parse(date.to_s) }
      }
    }
  end

  # COLLECTION-level aggregates: evaluated ONCE per request over the scoped,
  # filtered relation. Served by GET /api/{org}/tasks/computed?attributes=...
  #
  #   ?attributes=total_count,done_tasks_count            legacy comma list
  #   ?attributes[total_count]=                           bracket, no arguments
  #   ?attributes[count_by_status]=todo                   one declared parameter
  #   ?attributes[tasks_due_between][from]=a&...[to]=b    named parameters
  #
  # A bare GET /computed still returns everything the policy allows, minus any
  # attribute with a REQUIRED parameter — those are skipped silently, so adding
  # one here never breaks a client that asks for everything.
  def self.rhino_collection_computed_attributes
    {
      "total_count" => ->(scope, _user) { scope.count },
      "open_tasks_count" => ->(scope, _user) { scope.where.not(status: "done").count },
      "done_tasks_count" => ->(scope, _user) { scope.where(status: "done").count },
      "high_priority_count" => ->(scope, _user) { scope.where(priority: "high").count },
      "estimated_hours_total" => ->(scope, _user) { scope.sum(:estimated_hours).to_f },

      # Two required parameters: skipped by a bare /computed.
      "tasks_due_between" => {
        params: %i[from to],
        with: ->(scope, _user, from, to) { scope.where(due_date: from..to).count }
      },

      # One required parameter, so the bare-value form binds it:
      # ?attributes[count_by_status]=todo
      "count_by_status" => {
        params: [:status],
        with: ->(scope, _user, status) { scope.where(status: status).count }
      },

      # Every parameter optional: evaluated with no arguments by a bare
      # /computed, and narrowed when the client supplies a window.
      # 'only_high_priority' also shows "true"/"false" coercing to a real
      # boolean before the callable sees it.
      "windowed_task_count" => {
        params: %i[from to only_high_priority],
        optional: %i[from to only_high_priority],
        with: lambda { |scope, _user, from = nil, to = nil, only_high_priority = false|
          scope = scope.where("due_date >= ?", from) if from
          scope = scope.where("due_date <= ?", to) if to
          scope = scope.where(priority: "high") if only_high_priority == true
          scope.count
        }
      }
    }
  end

  belongs_to :project
  belongs_to :assignee, class_name: "User", optional: true
  has_many :comments, dependent: :destroy
  has_and_belongs_to_many :labels, join_table: :task_labels
end
